core/src/main/scala/scalaz/ImmutableArray.scala

```diff
@@ -158,15 +159,15 @@ object ImmutableArray extends ImmutableArrayInstances {
     def apply(idx: Int) = arr(idx)
 
     def length = arr.length
-    def toArray[B >: A : ClassTag] = arr.clone.asInstanceOf[Array[B]]
-    def copyToArray[B >: A](xs: Array[B], start: Int, len: Int): Unit = { arr.copyToArray(xs, start, len) }
+    def toArray[B: ClassTag](implicit ev: B >~> A): Array[B] = arr.clone.asInstanceOf[Array[B]]
+    def copyToArray[B](xs: Array[B], start: Int, len: Int)(implicit ev: B >~> A): Unit = { arr.asInstanceOf[Array[B]].copyToArray(xs, start, len) }
```

Comment:
Why is this `asInstanceOf` necessary? If it is indeed necessary, perhaps add a short comment explaining why.

Reply:
The reason is that provided an Array[A], copyToArray wants an Array[A] as input.  Any thing else and the compiler claims that it really wants an Array[Any], which is quite obviously bogus. 

```scala
[error] /Users/jacobbarber/projects/scalaz/core/src/main/scala/scalaz/ImmutableArray.scala:164:108: type mismatch;
[error]  found   : Array[B]
[error]  required: Array[Any]
[error] Note: B <: Any, but class Array is invariant in type T.
[error] You may wish to investigate a wildcard type such as `_ <: Any`. (SLS 3.2.10)
[error]     def copyToArray[B](xs: Array[B], start: Int, len: Int)(implicit ev: B >~> A): Unit = { arr.copyToArray(xs, start, len) }
[error]                                                                                                            ^
```

If I make `arr` (defined above) private to the package rather than private to `this`, then I can use the implicit evidence to convert `this` in to an `ImmutableArray1[B]`, and then perform the copy from there without a cast.  I'm not sure if it would be okay for `arr` to leak to other instances though.